From 2eb0d79cbeda9380b7d4e2d8ba38a73a78235452 Mon Sep 17 00:00:00 2001
From: Matti Kosola <matti.kosola@jolla.com>
Date: Thu, 6 Feb 2020 14:44:40 +0000
Subject: [PATCH] Disable docs

Signed-off-by: Matti Kosola <matti.kosola@jolla.com>
---
 Makefile.am | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/Makefile.am b/Makefile.am
index f37058f..10fd220 100644
--- a/Makefile.am
+++ b/Makefile.am
@@ -4,7 +4,7 @@
 
 AUTOMAKE_OPTIONS = 1.9 gnu dist-bzip2 dist-xz check-news
 
-SUBDIRS = libpam tests libpamc libpam_misc modules po doc examples xtests
+SUBDIRS = libpam tests libpamc libpam_misc modules po examples xtests
 
 CLEANFILES = *~
 
-- 
1.8.3-rc3

