From 49f1b264197b0b40f847a4c273e60bb349208dc8 Mon Sep 17 00:00:00 2001
From: Matti Kosola <matti.kosola@jolla.com>
Date: Thu, 6 Feb 2020 14:35:13 +0000
Subject: [PATCH] Disable environment man page

Signed-off-by: Matti Kosola <matti.kosola@jolla.com>
---
 modules/pam_env/Makefile.am | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/modules/pam_env/Makefile.am b/modules/pam_env/Makefile.am
index d6f081f..411405c 100644
--- a/modules/pam_env/Makefile.am
+++ b/modules/pam_env/Makefile.am
@@ -7,7 +7,7 @@ MAINTAINERCLEANFILES = $(MANS) README
 
 EXTRA_DIST = README pam_env.conf $(MANS) $(XMLS) tst-pam_env environment
 
-man_MANS = pam_env.conf.5 pam_env.8 environment.5
+man_MANS = pam_env.conf.5 pam_env.8
 
 XMLS = README.xml pam_env.conf.5.xml pam_env.8.xml
 
-- 
1.8.3-rc3

